### PR TITLE
Cleanup surface vertex points and reset vertex size 

### PR DESCRIPTION
**Pull request by lgu1234**
Pull request overview
---------------------
Address #6834.
1. Remove duplicated points of surface vertexes
2. Check colinearity
3. If colinearity, remove extra points
4. Resize vertex vectors
5. Provide a warning to show vertex reprocessing

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by lgu1234** [comment]
After submission of specific fix, I tried to use the debug version to run the program. E+ crashed due to mismatch of number of sides and vertex size from other surfaces. General fix is needed to prevent debug crash.

**Comment by lgu1234** [comment]
One more test file was uploaded in EnergyPlusDevSupport to test debug crash. In order to avoid debug crash, the code modification in CalcCoordinateTransformation is not enough to cover all and was removed. Instead, the code modification in CheckConvexity was added to have a general fix. The test files can run completion without fatal error and debug crash.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 1]
```diff
@@ -11019,7 +11019,7 @@ namespace SurfaceGeometry {
 
         DotSelfX23 = magnitude_squared(x23);
 
-        if (std::abs(DotSelfX23) <= .1e-6) {
+        if (DotSelfX23 <= .1e-6) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/SurfaceGeometry.cc:11022]: :+1: for efficiency.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 2]
```diff
@@ -12668,12 +12668,28 @@ namespace SurfaceGeometry {
...
+                    Ind = Ind - NSides + M - 1;
+                }
+                for (K = Ind; K <= NSides - 1; ++K) {
                     SurfaceTmp(SurfNum).Vertex(K - J + 1).x = SurfaceTmp(SurfNum).Vertex(K - J + 2).x;
                     SurfaceTmp(SurfNum).Vertex(K - J + 1).y = SurfaceTmp(SurfNum).Vertex(K - J + 2).y;
                     SurfaceTmp(SurfNum).Vertex(K - J + 1).z = SurfaceTmp(SurfNum).Vertex(K - J + 2).z;
                 }
             }
+            // remove duplicated points and resize Vertex
+            Array1D<Vector> OldVertex;
```
Comment (Myoldmopar) [inline src/EnergyPlus/SurfaceGeometry.cc:12681]: This method is only called once per simulation, right?

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 3]
```diff
@@ -3057,3 +3055,118 @@ TEST_F(EnergyPlusFixture, SurfaceGeometry_VertexNumberMismatchTest)
...
+    Surface(ThisSurf).Vertex(9).z = 0.0;
+
+    SurfaceTmp(ThisSurf) = Surface(ThisSurf);
+    CheckConvexity(ThisSurf, SurfaceTmp(ThisSurf).Sides);
+    Surface(ThisSurf) = SurfaceTmp(ThisSurf);
+    EXPECT_EQ(4, Surface(ThisSurf).Sides);
+    EXPECT_EQ(10.0, Surface(ThisSurf).Vertex(2).x);
+    EXPECT_EQ(8.0, Surface(ThisSurf).Vertex(2).y);
+    EXPECT_EQ(15.0, Surface(ThisSurf).Vertex(3).x);
+    EXPECT_EQ(8.0, Surface(ThisSurf).Vertex(3).y);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:3170]: This is a wonderful unit test! :+1: :+1:

**Review by Myoldmopar — COMMENTED** [review]
This seems just fine.  I haven't done anything locally but I will pull, build, and test it out tomorrow morning.  I won't mark it approve or reject yet because I want to confirm the context of that allocation/deallocation before doing so.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 2 above)
Comment (lgu1234) [reply src/EnergyPlus/SurfaceGeometry.cc:12681]: @Myoldmopar The method is called at each heat transfer surface with collinear points.

**Pull request merged**